sygnal/firebasepushkin.py

```diff
@@ -55,217 +76,233 @@ def __init__(self, name, sygnal, config):
             )
+        else:
+            return None
 
-        failed.extend(await self.send(message))
+    async def dispatch_notification(self, n, device, context):
+        log = NotificationLoggerAdapter(logger, {"request_id": context.request_id})
+
+        span_tags = {}
+        with self.sygnal.tracer.start_span(
+            "firebase_dispatch", tags=span_tags, child_of=context.opentracing_span
+        ) as span_parent:
+
+            dispatch_handler = self._map_event_dispatch_handler(n)
+            if dispatch_handler is None:
+                return []  # skipped
+
+            for retry_number in range(self.MAX_TRIES):
+                try:
+                    log.debug("Trying")
+
+                    span_tags = {"retry_num": retry_number}
+
+                    with self.sygnal.tracer.start_span(
+                        "firebase_dispatch_try", tags=span_tags, child_of=span_parent
+                    ) as span:
+                        return await dispatch_handler(device, span, log)
+                except TemporaryNotificationDispatchException as ex:
+                    retry_delay = self.RETRY_DELAY_BASE * (2 ** retry_number)
+                    if ex.custom_retry_delay is not None:
+                        retry_delay = ex.custom_retry_delay
+
+                    log.warning(
+                        "Temporary failure, will retry in %d seconds",
+                        retry_delay,
+                        exc_info=True,
+                    )
+                    span_parent.log_kv(
+                        {"event": "temporary_fail", "retrying_in": retry_delay}
+                    )
+                    if retry_number == self.MAX_TRIES - 1:
+                        raise NotificationDispatchException(
+                            "Retried too many times."
+                        ) from ex
+                    else:
+                        await twisted_sleep(
+                            retry_delay, twisted_reactor=self.sygnal.reactor
+                        )
 
-        return failed
+    @staticmethod
+    def _map_counts_unread(n):
+        return n.counts.unread or 0
 
-    async def send(self, message):
+    @staticmethod
+    def _map_android_priority(n):
+        return "normal" if n.prio == "low" else "high"
 
-        d = Deferred()
+    @staticmethod
+    def _map_ios_priority(n):
+        return "10" if n.prio == "high" else "5"
 
-        def done(success, result):
-            self.reactor.callFromThread(d.callback, result)
+    @staticmethod
+    def _message_body_from_notification(n, message_types):
+        if n.type != "m.room.message":
+            return ""
 
-        with SEND_TIME_HISTOGRAM.time():
-            self._pool.callInThreadWithCallback(
-                done, messaging.send_multicast, message, app=self._app
-            )
-            response = await d
+        from_display = ""
+        if n.room_name is not None and n.sender_display_name is not None:
```

Comment:
Why does `n.room_name` matter here?

Reply:
This is a special case which depends on preferences. We don't want to prepend the sender display name if the room is a direct chat (because the sender name is already in the title of the notification). We refactored this function to return the title and body with 599f1038bace53028f2671e4c6fae4f9c4d4c44c to make it more clear, but it remains a very specific behavior to our use case.

We are thinking about handling this (and some others cases) by making it possible to provide templates via the config and are open to suggestions from your side.

Reply:
Templates would be great (similar to how we do things in sydent/synapse). Might be out of scope for this PR though (as it's already quite huge).

Being able to do things based on whether a room is a direct chat or not may be more difficult however, unless the template supports `if` blocks.

Reply:
As this branch is merging into `jaywink/firebase-improvements` at the moment, feel free to leave this behaviour, and we can decide whether we want to implement templates when it comes time to merge that branch in.